Edit search term

VasilyShelkov opened the pull request:
fixes #54

@Winwardo ready for comments and merge...

VasilyShelkov committed:
Created Edit Dropdown

-created edit dropdown dumb component
-added createTwitterParamTypes builder +tested
-fixed searchActions addISearchTime with the new paramTypes builder + tested
-Added editSearchParamTypesSelection action

VasilyShelkov committed:
Moved the paramType builder to the reducer

VasilyShelkov committed:
Added toggleSearchTermParamTypeSelection

VasilyShelkov committed:
OnToggleClickAdded but bugged!

VasilyShelkov committed:
Merge remote-tracking branch 'refs/remotes/origin/master'

Conflicts:
	src/client/search/searchReducer.js
	src/client/search/searchReducer.spec.js

VasilyShelkov committed:
Finally got toggling termItemParamTypes working !

FYI don't use JQuery class selectors LOL

VasilyShelkov committed:
Added selection styling

Winwardo commented on public/css/customFonts.css:
```diff
@@ -29,6 +29,16 @@ i {
 }
 
+.ui.labeled.icon.button {
+  padding-left:1.5em!important;
```
![Codacy](https://www.codacy.com/assets/images/favicon.png) Issue found: [Use of !important](https://www.codacy.com/app/winwardo/solid-octo-disco/file/2283865473/issues/source?bid=3148782&fileBranchId=3188987#l33)

Winwardo replied:
![Codacy](https://www.codacy.com/assets/images/favicon.png) Issue found: [Use of !important](https://www.codacy.com/app/winwardo/solid-octo-disco/file/2286875896/issues/source?bid=3148782&fileBranchId=3188987#l33)

Winwardo replied:
![Codacy](https://www.codacy.com/assets/images/favicon.png) Issue found: [Use of !important](https://www.codacy.com/app/winwardo/solid-octo-disco/file/2288680902/issues/source?bid=3148782&fileBranchId=3188987#l33)

Winwardo replied:
![Codacy](https://www.codacy.com/assets/images/favicon.png) Issue found: [Use of !important](https://www.codacy.com/app/winwardo/solid-octo-disco/file/2288702796/issues/source?bid=3148782&fileBranchId=3188987#l33)

Winwardo commented on public/css/customFonts.css:
```diff
@@ -29,6 +29,16 @@ i {
 
+.ui.labeled.icon.button {
+  padding-left:1.5em!important;
+  padding-right:4.07142857em!important;
```
![Codacy](https://www.codacy.com/assets/images/favicon.png) Issue found: [Use of !important](https://www.codacy.com/app/winwardo/solid-octo-disco/file/2283865473/issues/source?bid=3148782&fileBranchId=3188987#l34)

Winwardo replied:
![Codacy](https://www.codacy.com/assets/images/favicon.png) Issue found: [Use of !important](https://www.codacy.com/app/winwardo/solid-octo-disco/file/2286875896/issues/source?bid=3148782&fileBranchId=3188987#l34)

Winwardo replied:
![Codacy](https://www.codacy.com/assets/images/favicon.png) Issue found: [Use of !important](https://www.codacy.com/app/winwardo/solid-octo-disco/file/2288680902/issues/source?bid=3148782&fileBranchId=3188987#l34)

Winwardo replied:
![Codacy](https://www.codacy.com/assets/images/favicon.png) Issue found: [Use of !important](https://www.codacy.com/app/winwardo/solid-octo-disco/file/2288702796/issues/source?bid=3148782&fileBranchId=3188987#l34)

VasilyShelkov committed:
Merged branch master into edit-search-term

Winwardo commented on src/client/search/TermItem.js:
```diff
@@ -0,0 +1,62 @@
+import React, { Component } from 'react';
+
+const TermItem = ({ onDeleteClick, onToggleParamTypeClick, query, source, paramTypes, id }) => (
+  <div id={`paramTypes${id}`} className="ui multiple compact dropdown labeled icon violet button"
```
Hmmmmmm, instead of appending the `id` to the id field, would a data-id={id} perhaps make more sense? Easier to search for and select programatically too

Winwardo commented on src/client/search/TermItem.js:
```diff
@@ -0,0 +1,62 @@
+
+const TermItem = ({ onDeleteClick, onToggleParamTypeClick, query, source, paramTypes, id }) => (
+  <div id={`paramTypes${id}`} className="ui multiple compact dropdown labeled icon violet button"
+    onClick={(ee) => ee.stopPropagation()}
```
Any reason for `ee` instead of `e` or `event`?

Winwardo commented on src/client/search/TermItem.js:
```diff
@@ -0,0 +1,62 @@
+      <i className={`${source} icon`}></i>
+      {query}
+    </span>
+    <i className="delete icon" onClick={() => {
```
onClick={onDeleteClick}

Winwardo commented on src/client/search/TermItem.js:
```diff
@@ -0,0 +1,62 @@
+
+class TermItemMenu extends Component {
+  componentDidMount() {
+    $(`#paramTypes${this.props.termId}`)
```
From above, you'd here search using this:
https://stackoverflow.com/questions/21756777/jquery-find-element-by-data-attribute-value

Feel free to keep it as an id rather than moving to data, I'd have gone for data.

Winwardo commented on src/client/search/searchReducer.js:
```diff
@@ -1,46 +1,61 @@
-  }
-};
\ No newline at end of file
+import {
```
I'm deeply unsettled by how this diff is entirely red and green, with no white :s

VasilyShelkov replied:
yer not sure whatt's happened here :S

Winwardo commented on src/shared/utilities.js:
```diff
@@ -82,13 +82,35 @@ export const fetchPost = (url, body) => (
+  name: 'author',
+  selected: paramTypesText.indexOf('author') > -1,
+  icon: getParamTypeIcon('author')
+}, {
```
This could very easily be refactored into:
[makeParamType('author'), makeParamType('hashtag'), ...]
where
makeParamType = (type) => ({name: type, selected: etc etc

Winwardo commented on src/shared/utilities.js:
```diff
@@ -82,13 +82,35 @@ export const fetchPost = (url, body) => (
- *
+ * Creates the paramTypes for a Twitter specific search term.
+ * @param array of strings representing search paramaterTypes
+ * @returns array of objects representing search terms with meta data
```
http://usejsdoc.org/tags-returns.html

Should be
@returns [{Object}]

Winwardo commented on src/shared/utilities.spec.js:
```diff
@@ -1,78 +1,107 @@
 import { should } from 'chai';
```
Was eslint killing you on this file? haha

VasilyShelkov committed:
Merged branch master into edit-search-term

VasilyShelkov committed:
Fixed the PR's comments

-changed dropdown html props from id=paramtypesid to data-id as suggested
-ee in the onclick became e
-removed unnecessary bracers onDeleteClick() and onToggleParamType
-refactored createTwitterParamTypes to have makeParamTypes to get rid of repeated code

VasilyShelkov committed:
fixed return comment param to [{Object}]

VasilyShelkov commented on src/shared/utilities.js:
```diff
@@ -82,13 +82,27 @@ export const fetchPost = (url, body) => (
+ */
+export const createTwitterParamTypes = (selectedParamTypes) =>
+  ['author', 'hashtag', 'keyword', 'mention']
+    .map((paramType) => makeParamType(selectedParamTypes, paramType));
```
@Winwardo I went a step further and mapped so didn't have to keep rewriting makeParamType.

Nice catch btw X)

Winwardo replied:
Was gonna suggest maybe you could just pass in a whole array but figured it was too much to ask for now haha

VasilyShelkov commented:
@Winwardo, ready for comments take two and merge...

hope you appreciated that createParamTypes refactor :P
